feat(next-auth): add getServerAuthSession

hnqg opened the pull request:
# Add getServerAuthSession helper and a Protected route using it

#### :white_check_mark: CHECKS

- [ x ] I reviewed linter warnings + errors, resolved formatting, types and other issues related to my work
- [ x ] The PR title follows the convention we established [conventional-commit](https://www.conventionalcommits.org/en/v1.0.0/)
- [ x ] I performed a functional test on my final commit

---

#### 📝 CHANGELOG

- I've created a function that abstracts the use of `unstable_getServerSession` from `next-auth`.
- To serve as an example how this function will going to be used, i've create a `Protected` route as well


💯

juliusmarminge commented:
Would you mind making the same change to next?

juliusmarminge reviewed (approved):
Great! Would be good with some refactoring to take advantage of this wrapper in the other places too:

https://github.com/t3-oss/create-t3-app/blob/93923741e4d44ff03b7e308ff5d1ea36b4ce3fcb/template/addons/trpc/auth-context.ts#L14-L15

https://github.com/t3-oss/create-t3-app/blob/93923741e4d44ff03b7e308ff5d1ea36b4ce3fcb/template/addons/trpc/auth-prisma-context.ts#L15-L16

https://github.com/t3-oss/create-t3-app/blob/93923741e4d44ff03b7e308ff5d1ea36b4ce3fcb/template/addons/next-auth/restricted.ts#L8

juliusmarminge commented:
@TheoBr Just to make sure, are we stepping on any toes when we disregarded protected routes in #122 (I think the main issue there was the JWT story)?

hnqg commented:
> Would you mind making the same change to next?

You mean the restricted route without tRPC?

https://github.com/t3-oss/create-t3-app/blob/main/template/addons/next-auth/restricted.ts

hnqg commented on template/addons/next-auth/protected-page.tsx:
```diff
@@ -0,0 +1,26 @@
+import { signOut } from "next-auth/react";
+import { getServerAuthSession } from "../server/common/get-server-auth-session";
+
+const ProtectedPage: NextPage = () => {
```
Definitely, gonna add some comments sugesting an usual flow on how to get the session data and use it on the page

hnqg replied:
Also will try to use the it in the places will suggest it. One little problem i came up with was with types, because the `getServerAuthSession` expects an object with `req` and `res` that are `GetServerSidePropsContext["req"]` and ``GetServerSidePropsContext["res"]``, but in some places (like the restricted endpoint withou tRPC) uses `req` and `res` as `NextApiRequest` and `NextApiResponse`.

I've just ignored it for the restricted enpoint because i think the `getServerAuthSession` was mainly proposed to be used in gSSP, but i will try in this files you pointed

juliusmarminge replied:
Oh wow are Next using different types for gSSP and API handlers...

Does it work for both? Like can we do something like
```ts
export const getServerAuthSession = async (
  ctx:
    | {
        req: GetServerSidePropsContext["req"];
        res: GetServerSidePropsContext["res"];
      }
    | { 
        req: NextApiRequest; 
        res: NextApiResponse;
      }
) => {
  return await unstable_getServerSession(ctx.req, ctx.res, nextAuthOptions);
};
```

hnqg replied:
Actually i've tried here and worked hahah my bad

Already commited!

juliusmarminge commented:
> > Would you mind making the same change to next?
> 
> 
> 
> You mean the restricted route without tRPC?
> 
> 
> 
> https://github.com/t3-oss/create-t3-app/blob/main/template/addons/next-auth/restricted.ts

I mean submitting the PR to the Next branch _as well_. Otherwise I'll do it after no problem

hnqg commented:
> > > Would you mind making the same change to next?
> > 
> > 
> > You mean the restricted route without tRPC?
> > https://github.com/t3-oss/create-t3-app/blob/main/template/addons/next-auth/restricted.ts
> 
> I mean submitting the PR to the Next branch _as well_. Otherwise I'll do it after no problem

Oh, okay! I will wait for this branch to be 100% ready to merge and will open another PR for the next branch

TheoBr reviewed (commented):
Concerned this is a lot of changes for a feature I don't see as super useful (and quite prescriptive). Many users might prefer to render a sign-in component on a "protected route" instead. The `as` bindings also scare me a bit as they weren't necessary before 🤔 

Not hard stanced either way just some thoughts

juliusmarminge commented:
> Not hard stanced either way just some thoughts

Yea I am good either way. It isn't a big deal to make myself when I end up needing it, but it is a nice to have to only import a single thing and not passing the nextAuthOptions around everywhere.

> `opts` argument is optional

Yea from what I can tell we should be good removing the optional here. This is the underlying typedef: 
```ts
export type NodeHTTPCreateContextFnOptions<TRequest, TResponse> = {
  req: TRequest;
  res: TResponse;
};
export type NodeHTTPCreateContextFn<
  TRouter extends AnyRouter,
  TRequest,
  TResponse,
> = (
  opts: NodeHTTPCreateContextFnOptions<TRequest, TResponse>,
) => inferRouterContext<TRouter> | Promise<inferRouterContext<TRouter>>;
```

What are your thoughts on this @nexxeln

hnqg commented:
sorry for sitting on this for so long, made the updates removing the type inferring on de arguments of `getServerAuthSession`.

also updated the `README` with a link to the protected page example. one thing I caught up on was that I couldn't access the `session` object as props in SSR pages, so I also removed that guide in the protected page and recommended accessing with the `useSession` hook as you would normally do.

Fedeorlandau commented:
Hi there!

First, thank you for your contribution.

I think this is too opinionated for create-t3-app. This protected page is using `getServerSideProps` which is not always what people want. You may want to protect your routes (for whatever reason) on the client side without having to involve a server call at all. Next/auth provides a lot of examples for doing this through middleware or even server-side on their official docs.

Just saying that it can be misleading.

Cheers,

juliusmarminge commented:
Continued in #366
